python/cudf/cudf/_lib/groupby.pyx

```diff
@@ -235,40 +226,33 @@ cdef class GroupBy:
                 c_agg_requests.push_back(
                     move(c_agg_request)
                 )
-
+                column_orders.append(i)
         if c_agg_requests.empty() and not allow_empty:
             raise DataError("All requested aggregations are unsupported.")
 
-        cdef pair[
-            unique_ptr[table],
-            vector[libcudf_groupby.aggregation_result]
-        ] c_result
-
         with nogil:
             c_result = move(
                 self.c_obj.get()[0].scan(
                     c_agg_requests
                 )
             )
 
-        grouped_keys, _ = data_from_unique_ptr(
-            move(c_result.first),
-            column_names=self.keys._column_names
+        grouped_keys = columns_from_unique_ptr(
+            move(c_result.first)
         )
 
-        result_data = ColumnAccessor(multiindex=True)
-        # Note: This loop relies on the included_aggregations dict being
-        # insertion ordered to map results to requested aggregations by index.
-        for i, col_name in enumerate(included_aggregations):
-            for j, agg_name in enumerate(included_aggregations[col_name]):
-                if callable(agg_name):
-                    agg_name = agg_name.__name__
-                result_data[(col_name, agg_name)] = (
-                    Column.from_unique_ptr(move(c_result.second[i].results[j]))
-                )
+        result_columns = [
+            [
+                Column.from_unique_ptr(
+                    move(c_result.second[i].results[j])
+                ) for j in range(c_result.second[i].results.size())
+            ] for i in range(c_result.second.size())
+        ]
+        result_columns_with_padding = [[]] * len(included_aggregations)
+        for i, cols in enumerate(result_columns):
+            result_columns_with_padding[column_orders[i]] = cols
```

Comment:
All the suggestions from above apply here too. It may even be worth defining a small `cdef` function for lines 244-253 to reuse in both places, but I'll leave that decision up to you.

Reply:
The `cdef` suggestion is applied.